usl_pipeline/usl_lib/tests/transformers/test_feature_raster_transformers.py

```diff
@@ -0,0 +1,186 @@
+import numpy
+from numpy import testing
+from shapely import geometry
+
+from usl_lib.shared import geo_data
+from usl_lib.transformers import feature_raster_transformers
+
+
+def bbox_polygon(x1: float, y1: float, x2: float, y2: float) -> geometry.Polygon:
+    return geometry.Polygon([(x1, y1), (x2, y1), (x2, y2), (x1, y2), (x1, y1)])
+
+
+def test_transform_to_feature_raster_layers():
+    infiltration_configuration = geo_data.DEFAULT_INFILTRATION_CONFIGURATION
+    max_hydraulic_conductivity = max(
+        [item.hydraulic_conductivity for item in infiltration_configuration.items]
+    )
+    max_wetting_front_suction_head = max(
+        [item.wetting_front_suction_head for item in infiltration_configuration.items]
+    )
+    max_effective_porosity = max(
+        [item.effective_porosity for item in infiltration_configuration.items]
+    )
+    max_effective_saturation = max(
+        [item.effective_saturation for item in infiltration_configuration.items]
+    )
+
+    header = geo_data.ElevationHeader(
+        col_count=3,
+        row_count=4,
+        x_ll_corner=0.0,
+        y_ll_corner=0.0,
+        cell_size=1.0,
+        nodata_value=-9999.0,
+    )
+    data = numpy.asarray(
+        [
+            [0.0, 1.0, 2.0],
+            [3.0, 4.0, -9999.0],
+            [6.0, 7.0, -9999.0],
+            [9.0, 10.0, 11.0],
+        ]
+    )
+    elevation = geo_data.Elevation(header=header, data=data)
+    boundaries = [
+        (bbox_polygon(1, 0, 3, 4), 1),  # exclude the first column (row:col=*:0)
+    ]
+    buildings = [
+        (bbox_polygon(0, 3, 3, 4), 1),  # this is the first row (row:col=0:*)
+    ]
+    green_areas = [
+        (bbox_polygon(0, 0, 3, 2), 1),  # these are last two rows (row:col=2,3:*)
+    ]
+    soil_classes = [
+        (bbox_polygon(1, 1, 2, 2), 3),  # row:col=2:1
+        (bbox_polygon(1, 0, 2, 1), 4),  # row:col=3:1
+        (bbox_polygon(2, 0, 3, 1), 11),  # row:col=3:2
+    ]
+    feature_matrix = feature_raster_transformers.transform_to_feature_raster_layers(
+        elevation,
+        boundaries,
+        buildings,
+        green_areas,
+        soil_classes,
+        infiltration_configuration,
+    ).tolist()
+
+    # Checking the first column (should be NODATA since outside the boundaries)
+    testing.assert_array_equal(
```

Comment:
it'd be good to check the dtype here, since it being float32s matters when we pass these into ML.
```python
testing.assert_array_equal(
  feature_matrix[0][0], 
  numpy.array([-9999.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], dtype=numpy.float32),
  strict=True,
)
```

could also check more of the array at once to avoid repetitive calls, e.g. drop the `tolist` call on `featute_matrix` above and do:

```python
numpy.testing.assert_array_equal(
  feature_matrix[:, 0, :],
  numpy.array(
      [[-9999.,     0.,     0.,     0.,     0.,     0.,     0.,     0.],
       [-9999.,     0.,     0.,     0.,     0.,     0.,     0.,     0.],
       [-9999.,     0.,     0.,     0.,     0.,     0.,     0.,     0.],
       [-9999.,     0.,     0.,     0.,     0.,     0.,     0.,     0.]],
      dtype=numpy.float32),
  strict=True
)

Reply:
Done